File: Google/Graph/444. Sequence Reconstruction.js
/**
 * @param {number[]} nums
 * @param {number[][]} sequences
 * @return {boolean}
 */
var sequenceReconstruction = function(nums, sequences) {
    if (nums.length <= 0) return false;

    const size = nums.length + 1;
    let adjacencyList = new Array(size);
    let indegrees = new Array(size).fill(0);
    if (buildGraph(sequences, adjacencyList, indegrees) !== nums.length) return false;

    let queue = [];
    for (let i = 0; i < indegrees.length; i++) {
    	if (indegrees[i] === 0) queue.push(i);
    }

    let index = 0;
    while (queue.length > 0) {
    	if (queue.length !== 1) return false;

    	const node = queue.pop();
    	if (node !== nums[index++]) return false;

    	for (const n of adjacencyList[node]) {
    		if (--indegrees[n] === 0) queue.push(n);
    	}
    }

    return index === nums.length;
};

var buildGraph = function(sequences, adjacencyList, indegrees) {
	let numberOfVertices = 0;

	for (let list of sequences) {
		for (let i = 0; i < list.length; i++) {
			const node = list[i];

			if (adjacencyList[node] === undefined) {
				adjacencyList[node] = [];
				numberOfVertices++;
			}

			if (i + 1 < list.length) {
				const next = list[i + 1];
				adjacencyList[node].push(next);
				indegrees[next]++;
			}
		}
	}

	return numberOfVertices;
};

File: Google/Graph/444. Sequence Reconstruction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetCode
{
    public class Program
    {
        public static void Main(string[] args)
        {
        }
    }

    public class Solution
    {
        public bool SequenceReconstruction(int[] nums, IList<IList<int>> sequences)
        {
            var size = nums.Length + 1;
            var graph = new IList<int>[size];
            var indegrees = new int[size];
            
            if (GenerateGraph(graph, indegrees, sequences) != nums.Length) return false;

            var topVertices = new Queue<int>();
            for (var i = 1; i < size; i++)
            {
                if (indegrees[i] == 0) topVertices.Enqueue(i);
            }

            var index = 0;
            while (topVertices.Count > 0)
            {
                if (topVertices.Count > 1) return false;

                var vertice = topVertices.Dequeue();

                if (nums[index++] != vertice) return false;

                for (var i = 0; i < graph[vertice].Count; i++)
                {
                    var node = graph[vertice][i];
                    if (--indegrees[node] == 0) topVertices.Enqueue(node);
                }
            }

            return index == nums.Length;
        }

        private int GenerateGraph(IList<int>[] graph, int[] indegrees, IList<IList<int>> sequences)
        {
            var numberOfVertices = 0;

            for (var i = 0; i < sequences.Count; i++)
            {
                for (var j = 0; j < sequences[i].Count; j++)
                {
                    var vertice = sequences[i][j];

                    if (graph[vertice] == null)
                    {
                        graph[vertice] = new List<int>();
                        numberOfVertices++;
                    }

                    if (j + 1 < sequences[i].Count)
                    {
                        graph[vertice].Add(sequences[i][j + 1]);
                        indegrees[sequences[i][j + 1]]++;
                    }
                }
            }

            return numberOfVertices;
        }
    }
}
